crypto/evp/evp_pkey.c

```diff
@@ -82,7 +89,20 @@ EVP_PKEY *EVP_PKCS82PKEY_ex(const PKCS8_PRIV_KEY_INFO *p8, OSSL_LIB_CTX *libctx,
     len = encoded_len;
     selection = EVP_PKEY_KEYPAIR | EVP_PKEY_KEY_PARAMETERS;
     dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "DER", "PrivateKeyInfo",
-                                         NULL, selection, libctx, propq);
+                                         keytype, selection, libctx, propq);
+
+    if (dctx != NULL && OSSL_DECODER_CTX_get_num_decoders(dctx) == 0) {
+        OSSL_DECODER_CTX_free(dctx);
+
+        /*
+         * This could happen if OBJ_obj2txt() returned a text OID and the
+         * decoder has not got that OID as an alias. We fall back to a NULL
+         * keytype
+         */
+        dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "DER", "PrivateKeyInfo",
+                                             NULL, selection, libctx, propq);
+    }
+
     if (dctx == NULL
         || !OSSL_DECODER_from_data(dctx, &p8_data, &len))
         /* try legacy */
```

Comment:
What is the legacy fallback needed here for?

Reply:
If there are no decoder because there are no keymgmt implementations, this falls back to try with engines and other legacy mechanisms.

Although, the condition seems a bit odd.  It should probably rather be the same as for the keytype = NULL fallback above, so something like this should probably be done I guess:

``` diff
-    if (dctx == NULL
-        || !OSSL_DECODER_from_data(dctx, &p8_data, &len))
+    if (dctx != NULL && OSSL_DECODER_CTX_get_num_decoders(ctx) == 0)
+        /* Error indicated by |pkey == NULL| */
+        (void)OSSL_DECODER_from_data(dctx, &p8_data, &len);
+    else
         /* try legacy */
```

Reply:
@mattcaswell, do you consider this relevant?

Reply:
> do you consider this relevant?

Not within the scope of this PR

Reply:
Fair enough